### PR TITLE
Replace Hallo with Draftail in table blocks

### PR DESCRIPTION
**Pull request by cwdavies**
The goal was to have replaced Hallo in rich text table blocks by the time Wagtail 2.8 was launched.
This change is necessary because rich text table cells no longer use Hallo for editing.

## Additions

Add support for Draftail in JobListingPageFormTests
- `cfgov/jobmanager/tests/models/test_pages.py`

Updates made to add Draftail support to the following files:
- `table-block.js`
- `package.json`
- `yarn.lock`
- `test.py`

## Removals

Hallo configuration to support rich text table cells is removed
Removed legacy wagtail admin rich text editor `Hallo` in `base.py`
Removed `hallotoolbar` from `table-block.css`
Updates made to remove Hallo support from the file:
- `table-block.css`
- `wagtail_hooks.py`

## Changes

- Converted from HTML to raw JS for Draftail initialization
- Updated modal close event listener when user clicks save button
- Changed `data-hallo-editor` to `richtext` in `insertRichTextDeleteControl`
- Replace `Hallo` with `Draftail` in table blocks in `cfgov/templates/wagtailadmin/js`
- Changed WAGTAILADMIN_RICH_TEXT_EDITORS to use DraftailRichText instead of HalloRichText in `test.py`

Renamed hallobutton to draftailButton in:
- html_editor.js
- ask_cfpb_tips.js

## How to test this PR

You can checkout this branch locally and use ./runserver.sh
or use the Docker container that this Pull Request creates to:
1. Open or create a wagtail page
2. Modify the rich text
3. Click to save the changes 
4. Verify the changes are correctly saved with no errors locked

## Notes

I feel the priority is to verify the functionality of clicking the save button that closes the modal.
It is very important that whatever is in the rich text editor is written to the table cell.
If that feature is NOT working, the Pull Request should not be merged until that
operation is confirmed to be working once the save button is clicked by the user.

## todos

- Verify draftail is properly setup in the model.
- Verify rich text buttons for bold, italics, etc. are added in this PR

## Checklist

- [x] PR has an informative and human-readable title
  - PR titles are used to generate the change log in [releases](../../releases); good ones make that easier to scan.
  - Consider prefixing, e.g., "Mega Menu: fix layout bug", or "Docs: Update Docker installation instructions".
- [x] Changes are limited to a single goal (no scope creep)
- [x] Code follows the standards laid out in the [CFPB development guidelines](https://github.com/cfpb/development)
- [x] Future todos are captured in comments and/or tickets
- [ ] Project documentation has been updated, potentially one or more of:
  - [This repo’s docs](https://cfpb.github.io/cfgov-refresh/) (edit the files in the `/docs` folder) – for basic, close-to-the-code docs on working with this repo
  - CFGOV/platform wiki on GHE – for internal CFPB developer guidance
  - CFPB/hubcap wiki on GHE – for internal CFPB design and content guidance

#### Browser testing

Visually tested in the following supported browsers:
- [ ] Firefox
- [ ] Chrome
- [ ] Safari
- [ ] Edge 18 (the last Edge prior to it switching to Chromium)
- [ ] Internet Explorer 11 and 8 (via emulation in 11's dev tools)
- [ ] Safari on iOS
- [ ] Chrome on Android

#### Accessibility

- [ ] Keyboard friendly (navigable with tab, space, enter, arrow keys, etc.)
- [ ] Screen reader friendly
- [ ] Does not introduce new errors or warnings in [WAVE](https://wave.webaim.org/extension/)

#### Other

- [ ] Is useable without CSS
- [ ] Is useable without JS
- [ ] Does not introduce new lint warnings
- [ ] Flexible from small to large screens

### CODE REVIEW

**Review by willbarton — DISMISSED** [review]
This doesn't seem quite there yet... 

Looking at the Coronavirus stimulus page, which has a table block on it with some rich text (http://localhost:8000/admin/pages/13837/edit/), it looks like there's a raw text box being included, and changes to the text don't actually get saved to the table block. In addition, we're missing a quite a lot of Draftail's editor features:

image

**Comment by willbarton** [comment]
Another thought I have is that it might be worth writing a functional test with Cypress for the existing Hallo table block rich text editor, and using that to test the Draftail editor.

File: cfgov/templates/wagtailadmin/js/table-block.js
Comment (cwdavies) [reply cfgov/templates/wagtailadmin/js/table-block.js]: Added missing label - I think this should be `class="data-draftail-input"`

**Comment by contolini** [comment]
Here are my notes from when @cwdavies and I were exploring draftail last week:

1. A user edits a cell by double clicking it. We [override](https://github.com/cfpb/cfgov-refresh/blob/2da707028df00acbe062177bd5799293ab8af5a9/cfgov/templates/wagtailadmin/js/table-block.js#L89) the default handsontable editing experience in order to do our fancy modal rich text editing experience. ([docs](https://handsontable.com/docs/7.4.2/tutorial-cell-editor.html))
1. The editing experience we provide involves [opening a modal](https://github.com/cfpb/cfgov-refresh/blob/2da707028df00acbe062177bd5799293ab8af5a9/cfgov/templates/wagtailadmin/js/table-block.js#L103) with an embedded rich text editor (formerly hallo but we're moving to draftail). We [initialize](https://github.com/cfpb/cfgov-refresh/blob/2da707028df00acbe062177bd5799293ab8af5a9/cfgov/templates/wagtailadmin/js/table-block.js#L549-L560) the rich text editor immediately after [opening the modal](https://github.com/cfpb/cfgov-refresh/blob/2da707028df00acbe062177bd5799293ab8af5a9/cfgov/templates/wagtailadmin/js/table-block.js#L104).
1. TODO: Ensure draftail is set up properly in the modal (we probably need to add [formatting options](https://www.draftail.org/docs/formatting-options) for bold, italics, etc.).
1. TODO: Update our [modal close event listener](https://github.com/cfpb/cfgov-refresh/blob/2da707028df00acbe062177bd5799293ab8af5a9/cfgov/templates/wagtailadmin/js/table-block.js#L106-L110) so that when the user clicks the save button (closing the modal), export whatever is in the modal's draftail rich text editor and write it to the handsontable table cell. We used to do this with hallo by just [calling `.html()` on the editor instance](https://github.com/cfpb/cfgov-refresh/blob/2da707028df00acbe062177bd5799293ab8af5a9/cfgov/templates/wagtailadmin/js/table-block.js#L107) but with draftail I think we have to take the editor's raw [ConstantState object](https://draftjs.org/docs/api-reference-content-state/) and use the `draft-convert` library to [convert it to html](https://www.draftail.org/docs/importing-and-exporting-html#html-export) which then gets injected into the cell using `instance.setDataAtCell()`.

> Another thought I have is that it might be worth writing a functional test with Cypress for the existing Hallo table block rich text editor, and using that to test the Draftail editor.

💯

**Comment by cwdavies** [comment]
@bartonw and @contolini I believe I found a solution to setting up draftail in the model. There are now specific rich text buttons for bold, italics, h2, h3, h4, ol, ul, etc. When you get a change would you run a `git pull` on the `draftables` branch and take a look. However, I have not been successful in exporting whatever is in the modal's draftail rich text editor and writing that to the handsontable table cell. It is though the table cell is read-only as the contents does not get changed when you click the save button. Should the contents be saved as JSON or HTML?

**Comment by alexhoffman617** [comment]
> This doesn't seem quite there yet...
> 
> Looking at the Coronavirus stimulus page, which has a table block on it with some rich text (http://localhost:8000/admin/pages/13837/edit/), it looks like there's a raw text box being included, and changes to the text don't actually get saved to the table block. In addition, we're missing a quite a lot of Draftail's editor features:
> 
> image

I believe all the functionality from hallo is now present

**Comment by alexhoffman617** [comment]
Remaining tasks:
- [ ] Fix bug where on editing a cell with a document, it shows up as a link in the editor
- [ ] Do not show raw text box in editor

Questions/Inconsistencies:
- [ ] In the Hallo editor there is a P tag button. Draftail considers a P tag button to be the unstyled/default state. I would think that the editor is fine without it but don't want to make assumptions.
- [ ] In the Hallo editor there is a remove link button. In draftail it seems like this is handled by just removing the link text. I think this is ok but don't want to assume.
- [ ] The order of the buttons is now different. I think this is fine but don't want to assume.
- [ ] There wasn't a test file corresponding for this file:
       - Is it ok to keep it untested?
       - Is there a django test covering the functionality that I'm unaware of?
       - Should we add unit tests?
       - Should we add cypress tests?

**Comment by Scotchester** [comment]
> @Scotchester can you take a look at this when you get a chance? Thanks!

Sorry I haven't had a chance to take a close look at this. Too many meetings going on. Tomorrow looks clear enough for me to take a deep dive, though!
 
> Remaining tasks:
> 
> * [ ]  Fix bug where on editing a cell with a document, it shows up as a link in the editor
>   - On looking at this more, I believe this was already happening in the previous version, just wasn't noticeable because previous editor did not show link vs document icon. Don't know if it is worth the time it takes to fix it since it doesn't affect user functionality at all, as links and documents function similarly

I'm not 100% sure what you mean by this, but I assume I'll get it once I see it in action. I'll take a look and let you know.

> Questions/Inconsistencies:
> 
> * [ ]  In the Hallo editor there is a P tag button. Draftail considers a P tag button to be the unstyled/default state. I would think that the editor is fine without it but don't want to make assumptions.
> * [ ]  In the Hallo editor there is a remove link button. In draftail it seems like this is handled by just removing the link text or clicking on it which gives the user the option to edit or remove. I think this is ok but don't want to assume.
> * [ ]  The order of the buttons is now different. I think this is fine but don't want to assume.

Yes, definitely good to use these as default in Draftail (no P button, no remove link button, and default button order).

> * [ ]  There wasn't a test file corresponding for this file:
>   - Is it ok to keep it untested?
>   - Is there a django test covering the functionality that I'm unaware of?
>   - Should we add unit tests?
>   - Should we add cypress tests?
>   Comment: Based on more investigation, it looks like the JS files in unprocessed are tested. The test files were not located where I expected them to be. I am going to create a test file and attempt to unit test table-block.js

Yes, I think we should create unit tests for this file. Should've had them before :) A simple Cypress functional test would probably also be good.

File: gulp/tasks/scripts.js
Comment (Scotchester) [inline gulp/tasks/scripts.js:230]:
```suggestion
```

File: gulp/tasks/scripts.js
Comment (Scotchester) [inline gulp/tasks/scripts.js:257]:
```suggestion
    gulp.parallel( 'scripts:modern' )
```

I'm sure this was helpful while you were iterating, but let's remove it before merging this PR, since working on admin scripts is a pretty rare use case.

Or, alternatively, could make a `scripts:watch:admin` variant.

**Review by Scotchester — COMMENTED** [review]
Outstanding work to all. I have a few minor questions and a bunch of trivial syntax suggestions, but I am happy to report that the functionality is working as expected! Looking forward to closing this out next week.

File: gulp/tasks/scripts.js
Comment (alexhoffman617) [reply gulp/tasks/scripts.js]: removed empty comments

File: cfgov/v1/wagtail_hooks.py
Comment (alexhoffman617) [reply cfgov/v1/wagtail_hooks.py:101]: verified that this is not needed

File: cfgov/v1/wagtail_hooks.py
Comment (alexhoffman617) [reply cfgov/v1/wagtail_hooks.py:124]: verified this is not needed

File: cfgov/v1/wagtail_hooks.py
Comment (alexhoffman617) [reply cfgov/v1/wagtail_hooks.py]: This is now pulled out

File: .gitignore
Comment (alexhoffman617) [reply .gitignore:10]: Dont know why we wouldnt need this we have a similar line:
cfgov/static_built/*
for the rest of the static built code. Maybe where the file outputs should be moved into cfgov/static_built?

Comment (Scotchester) [reply .gitignore:10]: Oops, I forgot to follow-up on this in the course of finishing my review on Friday. I'll take another look today.

File: cfgov/jobmanager/tests/models/test_pages.py
Comment (Scotchester) [reply cfgov/jobmanager/tests/models/test_pages.py:126]: @chosak cleared this up for me.

File: .gitignore
Comment (Scotchester) [reply .gitignore:10]: Ahahhh, this is because of the gymnastics we did to put the JS file back where it was originally after processing. I've just played around with it and determined that that was unnecessary. Commit coming to clean this up.

**Review by Scotchester — APPROVED** [review]
I think it's done!!

@alexhoffman617 Please pull down my recent commit and make sure it works for you, in case I did something stupid.

File: package.json
Comment (Scotchester) [inline package.json:59]: @alexhoffman617 This appears to be the cause of the Snyk error. Looks like a bad rebase or merge, since during the course of this PR we updated to highcharts 7.2.2, which is on the line below.

```suggestion
```

Comment (alexhoffman617) [reply package.json]: @Scotchester  got rid of the duplication, there still seems to be an error but it is no longer blocking the merge

**Pull request merged**
